kernel/thread.c

```diff
@@ -290,7 +290,45 @@ k_tid_t k_thread_create(struct k_thread *new_thread,
 	}
 	return new_thread;
 }
-#endif
+
+
+#ifdef CONFIG_USERSPACE
+u32_t _handler_k_thread_create(u32_t new_thread, u32_t stack_p,
+			       u32_t stack_size, u32_t entry,
+			       u32_t p1, u32_t arg6, void *ssf)
+{
+	struct _syscall_10_args *margs = (struct _syscall_10_args *)arg6;
+	k_thread_stack_t stack = (k_thread_stack_t)stack_p;
+	u32_t guard_size;
+
+	_SYSCALL_IS_OBJ(new_thread, K_OBJ_THREAD, 1, ssf);
+	_SYSCALL_MEMORY(margs, sizeof(*margs), 0, ssf);
+
+	/* Thread needs to have rw access to the entire stack region including
+	 * any guard areas
+	 */
+	guard_size = (u32_t)K_THREAD_STACK_BUFFER(stack) - (u32_t)stack;
+	_SYSCALL_MEMORY(stack, stack_size + guard_size, 1, ssf);
+
+	/* User threads may only create other user threads */
+	_SYSCALL_VERIFY(margs->arg9 & K_USER, ssf);
```

Comment:
There's a race condition here: if another thread is furiously flipping this bit, there's a possibility that a thread might be spawn with higher privileges from another user thread. I don't see thread creation as being something performance- or latency- sensitive, so I think it's a better idea to do this in a critical section where no other thread is allowed to run during thread creation.

Reply:
ah good catch. yeah for arguments passed in via struct, and not registers, we can have attacks like this.
I'll lock interrupts.